### DISCUSSION TITLE
[Zephyr support] Static allocation of Task stack

### DISCUSSION THREAD

**Discussion by sobkulir** [nasa/fprime#2026] [Q&A]
Hello,

I'm porting F' onto Zephyr RTOS and got it working with the baremetal scheduler. Now I started implementing the OSAL layer, but Threads in Zephyr take a statically allocated stack, while the `Os::Task::start(..., NATIVE_UINT_TYPE stackSize, ...)` takes the size of the stack and expects the implementation to allocate the memory dynamically. My question is, could you help me figure out how can the static stack allocation be supported in F'?

Here's an example Thread creation in Zephyr:
```cpp
#define MY_STACK_SIZE 500
#define MY_PRIORITY 5

// `my_stack_area` is allocated in the `__kstackmem` linker section.
K_THREAD_STACK_DEFINE(my_stack_area, MY_STACK_SIZE);

k_tid_t my_tid = k_thread_create(&my_thread_data, my_stack_area,
                                 K_THREAD_STACK_SIZEOF(my_stack_area),
                                 my_entry_point,
                                 NULL, NULL, NULL,
                                 MY_PRIORITY, 0, K_NO_WAIT);
```

My idea would be to:
1) Change the Autocoder to generate the `K_THREAD_STACK_DEFINE()` allocations when the CMAKE_SYSTEM_NAME is Zephyr
2) Make `Os::Task::start()` take an optional argument `const U8 *const stackAddr`.
3) Change the glue code calling the `Os::Task::start()` to pass the extra argument.

I'd be happy for any pointers to relevant sources or ideas!

**Comment by timcanham** [comment]
This is great! I've been eyeing Zephyr for a while! We would like to create an `fprime-community` port when you are done. I would recommend a PR from you with your #2 option. Then we can use it for other OS versions that need this.

**Reply by sobkulir** [reply]
I'll be happy to make make the port public once I get it to a decent shape with some tests. I was successfully avoiding making changes to the fprime repo until now, but with this issue, I'll have to fork.

Since I'm new to F', if you have a bit of time and could point me to relevant files, it would save me quite some digging.

Thank you for your quick answer!

**Comment by timcanham** [comment]
BTW, what board are you porting to?

**Reply by sobkulir** [reply]
We are using Nucleo h723zg (stm32h723zg) :)

**Comment by ThibFrgsGmz** [comment]
Hardly interested! I wanted to do it after porting FPrime To RTEMS on LEON3 with Renode ( if interested, repo is here https://github.com/ThibFrgsGmz/fprime-rtems-poc) .

But I failed and I am waiting for any update of the documentation about OSAL/buildsystem modification.

**Reply by sobkulir** [reply]
Ah cool! Do you have any parts of the OSAL implementation already by any chance? Yeah I spent one week in the buildsystem of F' to make it compile haha, but it'st well-documented and relatively traceable, especially compared to Zephyr.

**Reply by ThibFrgsGmz** [reply]
Negative I didn't manage to do anything in the end.

I'm completely lost with what to implement, and in CMake. 

PS:  I'm a bit tired of CMake. I'm only doing Bazel now - soon Buck2, and god that's much better.

**Comment by timcanham** [comment]
If you want a pattern for porting F Prime without forking it, you can take a look at our VxWorks adaptation:

https://github.com/fprime-community/fprime-vxworks

The basic concept is that a sub-repo (say for VxWorks or RTEMS or Zephyr) can contain the CMake toolchain files and the implementation of the OS classes for that platform:

CMake:

https://github.com/fprime-community/fprime-vxworks/tree/master/cmake/platform
https://github.com/fprime-community/fprime-vxworks/tree/master/cmake/toolchain

VxWorks OS adaptation files:

https://github.com/fprime-community/fprime-vxworks/tree/master/fprime-vxworks/VxWorks

Then if your project wants to use one of these adaptations, you can make it a sub-repo and point to it from the CMake file for the deployment that wants to run on that OS.

For instance, a local project here using VxWorks has this as the `settings.ini` content:

```ini
[fprime]
project_root: ../
framework_path: ../fprime
default_toolchain: gr740-vxworks7
ac_constants: ../config/AcConstants.ini
config_directory: ../config
library_locations: ../fprime-vxworks
environment_file: ../fprime-vxworks/cmake/env/VxWorks-GR740.ini
```

The `CMakeLists.txt` file has the following entry to include the subset of code in the OS-specific repo:

```
include("${FPRIME_PROJECT_ROOT}/fprime-vxworks/fprime-vxworks.cmake")
```

So, for Zephyr or RTEMS, you can create one of these sub-repos and the rest of the community can take advantage of it.

**Reply by sobkulir** [reply]
Thank you Tim!! I pushed my initial version without OSAL here: https://github.com/sobkulir/fprime-zephyr-app

Today, I started playing with the static stack allocation and ran into a few issues, e.g., Ref app from devel not building with the bleeding edge fpp-tools. I'm unsure if this is intended -- `requirements.txt` of devel contains fprime-tools v1.2.0, not bleeding edge so potentially yes. I added an issue on `fprime-community/fpp`, I'd be happy for a comment: https://github.com/fprime-community/fpp/issues/255

**Reply by ThibFrgsGmz** [reply]
Concerning my porting to RTEMS, my adventure ended with this bug (https://github.com/ThibFrgsGmz/fprime-rtems-poc/issues/4) that I could not solve. I was trying to create an active component or queue but I couldn't link the executable. I was inspired by the various projects fprime-shpinx/vxworks, fprime-freertos or fprime-threadx.

**Reply by timcanham** [reply]
We should be able to get the RTEMS build working. Can you put your latest error log on the git issue you reference above?

**Comment by sobkulir** [comment]
I got the Tasks working in a bit of a hacky way just to move forward (https://github.com/fprime-community/fpp/compare/main...sobkulir:fpp:static_stack).

Today I implemented the `Os/Queue.hpp` API ([code](https://github.com/sobkulir/fprime-zephyr-app/blob/main/fprime/fprime-zephyr/Os/Queue.cpp)), however, I ran into a few issues:
1) How important is the message priority for standard components? I don't see a trivial way to support it in Zephyr and some other ports also ignore it.
2) Do standard components need message sizes of more than a fixed size, say 256B? Since Zephyr does not support variable-sized messages in its [Message Queues](https://docs.zephyrproject.org/3.1.0/kernel/services/data_passing/message_queues.html), I'm passing the size as the first byte. However, this also means that an extra buffer of 256B needs to be allocated on the thread stack and incur extra mem copy. I got the idea from [ThreadX port](https://github.com/fprime-community/fprime-azure-threadx/blob/fprime-azure-threadx-cmjl-14-01-2022-v00/fprime-nucleo/NucleoOs/Queue.cpp#L192)

I know that I'd be able to answer these types of questions if I had first dived into F' before doing the port, however, we decided to first see if we can even reasonably port it before using it. I'd be happy for your insights!

**Comment by LeStarch** [comment]
Following up on our discussion here!  @ethancheez has built on this work to get a running LED blink example (rate groups, active components with threads, etc) running on the teensy 41 board.  Take a look: https://github.com/fprime-community/fprime-zephyr-led-blinker

**Reply by LeStarch** [reply]
Note: we got around the static-thread allocation by leveraging Zephyr's experimental dynamic thread allocation instead.
